### ISSUE TITLE
Accessibility audit

### ISSUE THREAD

**Issue by patrycjanie** [PAYONE-GmbH/magento-2#594]
Hi,
Thanks for handlig my previous issue so quickly, I appreciate that a lot. I have one more issue to be discussed. During a11y audit run we noticed a few elements that are not compliant with WCAG 2.1 rules. My project is using credit card payment method and I prepared fix based on [this](https://github.com/PAYONE-GmbH/magento-2/blob/master/view/frontend/web/template/payment/creditcard.html):

```<!--
/**
 * PAYONE Magento 2 Connector is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PAYONE Magento 2 Connector is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with PAYONE Magento 2 Connector. If not, see <http://www.gnu.org/licenses/>.
 *
 * PHP version 5
 *
 * @category  Payone
 * @package   Payone_Magento2_Plugin
 * @author    FATCHIP GmbH <support@fatchip.de>
 * @copyright 2003 - 2016 Payone GmbH
 * @license   <http://www.gnu.org/licenses/> GNU Lesser General Public License
 * @link      http://www.payone.de
 */
-->
<div class="payment-method" data-bind="css: {'_active': (getCode() == isChecked())}">
    <div class="payment-method-title field choice">
        <input type="radio"
               name="payment[method]"
               class="radio"
               data-bind="attr: {'id': getCode()}, value: getCode(), checked: isChecked, click: selectPaymentMethod, visible: isRadioButtonVisible()"/>
        <label data-bind="attr: {'for': getCode()}" class="label"><span data-bind="text: getTitle()"></span></label>
    </div>

    <div class="payment-method-content">
        <!-- ko foreach: getRegion('messages') -->
        <!-- ko template: getTemplate() --><!-- /ko -->
        <!--/ko-->
        <div class="payment-method-billing-address">
            <!-- ko foreach: $parent.getRegion(getBillingAddressFormName()) -->
            <!-- ko template: getTemplate() --><!-- /ko -->
            <!--/ko-->
        </div>

        <form class="creditcard_form" id="creditcard_form" action="#" method="post" data-bind="attr: {id: getCode() + '_creditcard_form'}">
            <input type="hidden" name="payment[cc_pseudocardpan]" value=""
                   data-bind="attr: {autocomplete: off, id: getCode() + '_pseudocardpan'},
                              value: pseudocardpan">

            <fieldset data-bind="attr: {class: 'fieldset payment items ccard ' + getCode(), id: 'payone_form_cc'}">
            <legend class="visually-hidden" data-bind="i18n: 'Credit Card'"></legend>
                <!-- ko if: useSaveDataMode() -->
                <div data-bind="foreach: getSavedPaymentData()">
                    <div class="payment-method">
                        <div class="payment-method-title field choice">
                            <input data-bind="attr: {id: $parent.getCode() + '_data_' + $data.id, name: $parent.getCode() + '_saved_data'}, value: $data.payment_data.cardpan, event: { change: $parent.handleNewDataVisibility }" type="radio" class="radio"/>
                            <label data-bind="attr: {for: $parent.getCode() + '_data_' + $data.id}" class="label">
                                <img data-bind="attr: {src: 'https://cdn.pay1.de/cc/' + $data.payment_data.cardtype.toLowerCase() + '/s/default.png'}"/>
                                <!-- ko i18n: 'ends with'--><!-- /ko -->
                                <span data-bind="text: '-' + $data.payment_data.masked.substring(($data.payment_data.masked.length-4), $data.payment_data.masked.length)"></span>
                                <span data-bind="text: $data.payment_data.cardholder" class="cardholder"></span>
                                <span data-bind="text: $data.payment_data.cardexpiredate.substring(2, 4) + '/20' + $data.payment_data.cardexpiredate.substring(0, 2)"></span>
                            </label>
                        </div>
                    </div>
                </div>
                <div class="payment-method">
                    <div class="payment-method-title field choice">
                        <input data-bind="attr: {id: getCode() + '_new_data', name: getCode() + '_saved_data'}, event: { change: handleNewDataVisibility }" type="radio" class="radio" value="new"/>
                        <label data-bind="attr: {for: getCode() + '_new_data'}" class="label"><!-- ko i18n: 'Add new creditcard'--><!-- /ko --></label>
                    </div>
                </div>
                <!--/ko-->
                <div data-bind="attr: {id: getCode() + '_new_data_container'}" class="fieldset ccard">
                    <!-- ko if: isAutoCardtypeDetectionEnabled() -->
                    <input name="payment[cc_type]" type="hidden" value="" data-bind="attr: {id: getCode() + '_credit_card_type', value: getCreditcardTypes()[0].id, 'data-container': getCode() + '_credit_card_type'}" />
                    <!--/ko-->
                    <!-- ko if: !isAutoCardtypeDetectionEnabled() -->
                    <div class="field">
                        <label data-bind="attr: {for: getCode() + '_credit_card_type'}" class="label">
                            <span><!-- ko i18n: 'Credit Card Type'--><!-- /ko --></span>
                        </label>
                        <div class='control'>
                            <select style="width:200px;" name="payment[cc_type]" class="select"
                                    data-bind=" attr: {id: getCode() + '_credit_card_type', 'data-container': getCode() + '_credit_card_type', 'autocomplete': 'off'},
                                            mageInit: {creditCardType:{creditCardTypeContainer:'#' + getCode() + '_credit_card_type_ss_div'}},
                                            enable: true,
                                            options: getCreditcardTypes(),
                                            optionsValue: 'id',
                                            optionsText: 'title',
                                            optionsCaption: $t('--Please Select--')">
                            </select>
                        </div>
                    </div>
                    <!--/ko-->

                    <div class="field number required">
                        <label data-bind="attr: {for: getCode() + '_cardholder'}" class="label">
                            <span><!-- ko i18n: 'Cardholder'--><!-- /ko --></span>
                        </label>
                        <div class="control">
                            <input type="text" name="payment[cc_cardholder]" class="input-text" value=""
                                   data-bind="attr: {
                                        autocomplete: 'off',
                                        id: getCode() + '_cardholder',
                                        'data-container': getCode() + '-cardholder'},
                                  value: cardholder,
                                  valueUpdate: 'keyup' "/>
                        </div>
                    </div>

                    <div class="field number required">
                        <label class="label" for="cardpan">
                            <span><!-- ko i18n: 'Credit Card Number'--><!-- /ko --></span>
                            <!-- ko if: isAutoCardtypeDetectionEnabled() -->
                            <!-- ko foreach: getCreditcardTypes() -->
                            <img class="cc-icon" data-bind="attr: {id: $parent.getCode() + '_cc_icon_' + id.toLowerCase(), src: 'https://cdn.pay1.de/cc/' + id.toLowerCase() + '/s/default.png'}"/>
                            <!--/ko-->
                            <!--/ko-->
                        </label>
                        <div class="control">
                            <span id="cardpan" class="inputIframe"></span>
                        </div>
                    </div>

                    <div class="field date required" data-bind="attr: {id: getCode() + '_cc_type_exp_div'}">
                        <label class="label" for="cardexpiremonth">
                            <span><!-- ko i18n: 'Expiration Date'--><!-- /ko --></span>
                        </label>
                        <div class="control">
                            <div class="fields group group-2">
                                <div class="field no-label month">
                                    <div class="control">
                                        <span id="cardexpiremonth"></span>
                                    </div>
                                </div>
                                <div class="field no-label year">
                                    <div class="control">
                                        <span id="cardexpireyear"></span>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>

                    <!-- ko if: showCvc() -->
                    <div class="field cvv required" data-bind="attr: {id: getCode() + '_cc_type_cvv_div'}">
                        <label class="label" for="cardcvc2">
                            <span><!-- ko i18n: 'Card Verification Number'--><!-- /ko --></span>
                        </label>
                        <div class="control">
                            <span id="cardcvc2" class="inputIframe"></span>
                        </div>
                    </div>
                    <!--/ko-->

                    <!-- ko if: isSaveDataEnabled() -->
                    <input type="checkbox" data-bind="attr: {id: getCode() + '_save_data_checkbox'}, checked: saveData" />
                    <label data-bind="attr: {for: getCode() + '_save_data_checkbox'}" class="required" style="float:none;">
                        <!-- ko i18n: 'Save the payment data for future use.'--><!-- /ko --><br>
                    </label>
                    <!--/ko-->
                </div>
            </fieldset>
        </form>
        <div data-bind="html: handleIframes()"></div>
        <span data-bind="html: markDefaultSavedPayment()"></span>

        <p data-bind="html: getInstructions()"></p>
        <div class="checkout-agreements-block">
            <!-- ko foreach: $parent.getRegion('before-place-order') -->
            <!-- ko template: getTemplate() --><!-- /ko -->
            <!--/ko-->
        </div>
        <div data-role="checkout-payone-boni-agreement">
            <div class="checkout-payone-boni-agreement">
                <!-- ko if: (isAgreementVisible()) -->
                    <!-- ko if: (canShowPaymentHintText()) -->
                        <div>
                            <strong><span data-bind="html: getPaymentHintText()"></span></strong>
                        </div>
                    <!-- /ko -->
                    <!-- ko if: (canShowAgreementMessage()) -->
                        <div>
                            <input type="checkbox"
                                   data-bind="attr: {
                                                        'id': 'payone_boni_agreement_' + getCode(),
                                                        'name': 'payone_boni_agreement',
                                                        'value': '1'}" />
                            <label data-bind="attr: {'for': 'payone_boni_agreement_' + getCode()}">
                                <span data-bind="html: getAgreementMessage()"></span>
                            </label>
                        </div>
                    <!-- /ko -->
                <!-- /ko -->
            </div>
        </div>
        <div class="actions-toolbar">
            <div class="primary">
                <button class="action primary checkout"
                        type="submit"
                        data-bind="
                        click: handleCreditcardPayment,
                        attr: {title: $t('Place Order')},
                        enable: (getCode() == isChecked()),
                        css: {disabled: !isPlaceOrderActionAllowed()}
                        "
                        disabled>
                    <span data-bind="i18n: 'Place Order'"></span>
                </button>
            </div>
        </div>
    </div>
</div>
```
As you can see based on above, the issues are mostly connected to Success Criterion 4.1.2 Name, Role, Value
- The `<fieldset>` has no `<legend>` element.
- The `<label>` does not have a for attribute to associate it with a form control.
- The form control uses multiple naming techniques.
-  The form field is missing a valid autocomplete="..." attribute.

Also we have 2 issues related to iframes that are used in form fields:
- lang not defined on html Description: The page lacks a lang="..." attribute on the <html> element.
- credit card iframes are missing title
but I don't know if you have control over that.

Would you consider running a11y check on the your templates and adjust them to be comliant with WCAG 2.1?

**Comment by janteuber** [comment]
Hello @patrycjanie 

Thank you very much for the helpful hints. We have implemented them and will incorporate them into our plugin.

**Comment by patrycjanie** [comment]
thanks a lot!
